src/peer_loop.rs

```diff
@@ -1632,15 +1632,31 @@ impl PeerLoopHandler {
                 peer.send(PeerMessage::PeerListRequest).await?;
                 Ok(KEEP_CONNECTION_ALIVE)
             }
-            MainToPeerTask::Disconnect(target_socket_addr) => {
+            MainToPeerTask::Disconnect(peer_address) => {
                 log_slow_scope!(fn_name!() + "::MainToPeerTask::Disconnect");
 
-                // Disconnect from this peer if its address matches that which the main
-                // task requested to disconnect from.
-                Ok(target_socket_addr == self.peer_address)
+                // Only disconnect from the peer the main task requested a disconnect for.
+                if peer_address != self.peer_address {
+                    return Ok(KEEP_CONNECTION_ALIVE);
+                }
+                self.global_state_lock
+                    .lock_guard_mut()
+                    .await
+                    .net
+                    .register_peer_disconnection(peer_address, SystemTime::now());
+
+                Ok(DISCONNECT_CONNECTION)
+            }
+            MainToPeerTask::DisconnectAll() => {
+                // Disconnect from this peer, no matter what.
+                self.global_state_lock
+                    .lock_guard_mut()
+                    .await
+                    .net
+                    .register_peer_disconnection(self.peer_address, SystemTime::now());
```

Comment:
I would *only* add to this hash map in the scenario where you disconnect from a specified peer. But that's not a strong preference. Consider your opinion, then go with that.